File: flutter_twilio_conversations_web/lib/methods/listeners/channel_listener.dart
import 'dart:async';
import 'dart:js_util';
import 'package:flutter_twilio_conversations/flutter_twilio_conversations.dart';
import 'package:flutter_twilio_conversations_web/interop/classes/client.dart'
    as TwilioChatClient;
import 'package:flutter_twilio_conversations_web/interop/classes/channel.dart'
    as TwilioClientConversation;

import 'package:flutter_twilio_conversations_web/methods/listeners/base_listener.dart';
import 'package:flutter_twilio_conversations_web/methods/mapper.dart';
import 'package:js/js.dart';

// TODO implement this listener
// messageAdded
// messageRemoved
// messageUpdated
// participantJoined
// participantLeft
// participantUpdated
// pushNotification
// typingEnded
// typingStarted
class ChannelEventListener extends BaseListener {
  final TwilioClientConversation.TwilioConversationsChannel _channel;
  final StreamController<Map<String, dynamic>> _channelStreamController;

  ChannelEventListener(this._channel, this._channelStreamController) {}

  void addListeners() {
    debug('Adding chatClientEventListeners for ${_channel.sid}');
    _on('messageAdded', messageAdded);
    _on('messageUpdated', messageUpdated);
  }

  void _on(String eventName, Function eventHandler) => _channel.on(
        eventName,
        allowInterop(eventHandler),
      );

  void _off(String eventName, Function eventHandler) => _channel.off(
        eventName,
        allowInterop(eventHandler),
      );

  messageAdded(dynamic message) {
    debug('Message Added Channel Event');
    sendEvent("messageAdded", Mapper.messageToMap(message));
  }

  messageUpdated(dynamic data) {
    debug('Message Updated Channel Event');

    sendEvent("messageUpdated", {
      "message": Mapper.messageToMap(data.message),
      "reason": {"type": "message", "value": data.reason.toString()}
    });
  }

  // override fun onMessageDeleted(message: Message) {
  //     Log.d("TwilioInfo", "ChannelListener.onMessageDeleted => messageSid = ${message.sid}")
  //     sendEvent("messageDeleted", mapOf("message" to Mapper.messageToMap(message)))
  // }

  // override fun onParticipantAdded(member: Participant) {
  //     Log.d("TwilioInfo", "ChannelListener.onMemberAdded => memberSid = ${member.sid}")
  //     sendEvent("memberAdded", mapOf("member" to Mapper.memberToMap(member)))
  // }

  // override fun onParticipantUpdated(member: Participant, reason: Participant.UpdateReason) {
  //     Log.d("TwilioInfo", "ChannelListener.onMemberUpdated => memberSid = ${member.sid}, reason = $reason")
  //     sendEvent("memberUpdated", mapOf(
  //             "member" to Mapper.memberToMap(member),
  //             "reason" to mapOf(
  //                     "type" to "member",
  //                     "value" to reason.toString()
  //             )
  //     ))
  // }

  // override fun onParticipantDeleted(member: Participant) {
  //     Log.d("TwilioInfo", "ChannelListener.onMemberDeleted => memberSid = ${member.sid}")
  //     sendEvent("memberDeleted", mapOf("member" to Mapper.memberToMap(member)))
  // }

  // override fun onTypingStarted(channel: Conversation, member: Participant) {
  //     Log.d("TwilioInfo", "ChannelListener.onTypingStarted => channelSid = ${channel.sid}, memberSid = ${member.sid}")
  //     sendEvent("typingStarted", mapOf("channel" to Mapper.channelToMap(pluginInstance, channel), "member" to Mapper.memberToMap(member)))
  // }

  // override fun onTypingEnded(channel: Conversation, member: Participant) {
  //     Log.d("TwilioInfo", "ChannelListener.onTypingEnded => channelSid = ${channel.sid}, memberSid = ${member.sid}")
  //     sendEvent("typingEnded", mapOf("channel" to Mapper.channelToMap(pluginInstance, channel), "member" to Mapper.memberToMap(member)))
  // }

  // override fun onSynchronizationChanged(channel: Conversation) {
  //     Log.d("TwilioInfo", "ChannelListener.onSynchronizationChanged => channelSid = ${channel.sid}")
  //     sendEvent("synchronizationChanged", mapOf("channel" to Mapper.channelToMap(pluginInstance, channel)))
  // }

  // private fun sendEvent(name: String, data: Any?, e: ErrorInfo? = null) {
  //     val eventData = mapOf("name" to name, "data" to data, "error" to Mapper.errorInfoToMap(e))
  //     events.success(eventData)
  // }

  sendEvent(String name, dynamic data, {ErrorInfo? e}) {
    final eventData = {
      "name": name,
      "data": data,
      "error": Mapper.errorInfoToMap(e),
    };
    _channelStreamController.add(eventData);
  }
}

File: flutter_twilio_conversations/ios/Classes/Listeners/ChannelListener.swift
import Flutter
import TwilioConversationsClient

public class ChannelListener: NSObject, TCHConversationDelegate {
    let events: FlutterEventSink

    init(_ events: @escaping FlutterEventSink) {
        self.events = events
    }

    // onMessageAdded
    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, messageAdded message: TCHMessage) {
        SwiftTwilioConversationsPlugin.debug(
            "ChannelListener.onMessageAdded => messageSid = \(String(describing: message.sid))")
        sendEvent("messageAdded", data: [
            "message": Mapper.messageToDict(message, channelSid: conversation.sid)
        ])
    }

    // onMessageUpdated
    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, message: TCHMessage, updated: TCHMessageUpdate) {
        SwiftTwilioConversationsPlugin.debug(
            "ChannelListener.onMessageUpdated => messageSid = \(String(describing: message.sid)), " +
            "updated = \(String(describing: updated))")
        sendEvent("messageUpdated", data: [
            "message": Mapper.messageToDict(message, channelSid: conversation.sid),
            "reason": [
                "type": "message",
                "value": Mapper.messageUpdateToString(updated)
            ]
        ])
    }

    // onMessageDeleted
    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, messageDeleted message: TCHMessage) {
        SwiftTwilioConversationsPlugin.debug(
            "ChannelListener.onMessageDeleted => messageSid = \(String(describing: message.sid))")
        sendEvent("messageDeleted", data: [
            "message": Mapper.messageToDict(message, channelSid: conversation.sid)
        ])
    }

    // onMemberAdded
    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, participantJoined participant: TCHParticipant) {
        SwiftTwilioConversationsPlugin.debug(
            "ChannelListener.onMemberAdded => memberSid = \(String(describing: participant.sid))")
        sendEvent("memberAdded", data: [
            "member": Mapper.memberToDict(participant, channelSid: conversation.sid) as Any
        ])
    }

    // onMemberUpdated
    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, participant: TCHParticipant, updated: TCHParticipantUpdate) {
        SwiftTwilioConversationsPlugin.debug(
            "ChannelListener.onMemberUpdated => memberSid = \(String(describing: participant.sid)), " +
            "updated = \(String(describing: updated))")
        sendEvent("memberUpdated", data: [
            "member": Mapper.memberToDict(participant, channelSid: conversation.sid) as Any,
            "reason": [
                "type": "member",
                "value": Mapper.memberUpdateToString(updated)
            ]
        ])
    }

    // onMemberDeleted
    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation, participantLeft participant: TCHParticipant) {
        SwiftTwilioConversationsPlugin.debug(
            "ChannelListener.onMemberDeleted => memberSid = \(String(describing: participant.sid))")
        sendEvent("memberDeleted", data: [
            "member": Mapper.memberToDict(participant, channelSid: conversation.sid) as Any
        ])
    }

    // onTypingStarted
    public func conversationsClient(_ client: TwilioConversationsClient, typingStartedOn conversation: TCHConversation, participant: TCHParticipant) {
        SwiftTwilioConversationsPlugin.debug(
            "ChannelListener.onTypingStarted => channelSid = \(String(describing: conversation.sid)), " +
            "memberSid = \(String(describing: participant.sid))")
        sendEvent("typingStarted", data: [
            "channel": Mapper.channelToDict(conversation) as Any,
            "member": Mapper.memberToDict(participant, channelSid: conversation.sid) as Any
        ])
    }

    // onTypingEnded
    public func conversationsClient(_ client: TwilioConversationsClient, typingEndedOn conversation: TCHConversation, participant: TCHParticipant)  {
        SwiftTwilioConversationsPlugin.debug(
            "ChannelListener.onTypingEnded => channelSid = \(String(describing: conversation.sid)), " +
            "memberSid = \(String(describing: participant.sid))")
        sendEvent("typingEnded", data: [
            "channel": Mapper.channelToDict(conversation) as Any,
            "member": Mapper.memberToDict(participant, channelSid: conversation.sid) as Any
        ])
    }

    // onSynchronizationChanged
    public func conversationsClient(_ client: TwilioConversationsClient, conversation: TCHConversation,
        synchronizationStatusUpdated status: TCHConversationSynchronizationStatus) {
        SwiftTwilioConversationsPlugin.debug(
            "ChannelListener.onSynchronizationChanged => channelSid = \(String(describing: conversation.sid))")
        sendEvent("synchronizationChanged", data: [
            "channel": Mapper.channelToDict(conversation) as Any
        ])
    }

    private func sendEvent(_ name: String, data: [String: Any]? = nil, error: Error? = nil) {
        let eventData = [
            "name": name,
            "data": data,
            "error": Mapper.errorToDict(error)
            ] as [String: Any?]

        events(eventData)
    }
}
